File: 0152-maximum-product-subarray/0152-maximum-product-subarray.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int res = nums[0];
        int curr_min = 1;
        int curr_max = 1;
        
        for (int i = 0; i < nums.size(); i++) {
            int n = nums[i];
            int temp = curr_max * n;
            curr_max = max(max(n * curr_max, n * curr_min), n);
            curr_min = min(min(temp, n * curr_min), n);
            res = max(res, curr_max);
        }
        return res;
    }
};

File: 0152-maximum-product-subarray/0152-maximum-product-subarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        res = nums[0]
        curr_max = 1
        curr_min = 1
        
        for n in nums:
            temp = curr_max * n
            curr_max = max(curr_max * n , curr_min * n, n)
            curr_min = min(curr_min * n, temp, n)
            res = max(res, curr_max)
            
        return res